### DISCUSSION TITLE
如何解决同个仓库下多个项目共有代码问题？

### DISCUSSION THREAD

**Discussion by xmsz** [midwayjs/midway#1415] [Q&A]
背景
```
app1
app2
...
shared

```
- app想要直接引用shared的文件

不知道这么做比较好

我想到的几个方向是
1. 使用软连接，用系统层面去解决，但是一听就很麻烦
2. 使用构建工具，这样开发就需要多开一个服务，也很麻烦
3. npm link 但这个只适应包，更新也麻烦，而且只能开发环境搞，当然可以升级成monorepo
4. submodule/subtree，这种适合跨项目的，单个项目也太繁琐了
5. Typescript的Project References 还没尝试过，大概[看了下](https://github.com/RyanCavanaugh/project-references-demo)，应该也是全局构建，不太合适
6. pnpm/yarn link 解决了npm的更新的问题，这个不错，可以代替简单的monorepo需求

大概总结就是

- 如果省步骤和麻烦，那【软连接】不错
因为增加和删除都很简单，文件还是实时同步的，而且不需要把共享代码变成一个包，很方便

- 如果只是文件共享，不觉得配置麻烦的话，那【构建工具】不错
好处就是没有什么其他概念，就是只是纯粹的编译到不同目录而已

- 如果是多项目共享，并且也是文件的话，那【subtree】不错
方便，无负担

- 如果是包共享，简单需求，那【yarn link】不错
但只适合开发 因为和软连接一样 docker构建不了

- 如果是包共享，复杂需求，例如需要共同发布的，那肯定【lerna】不错
真正解决方案

大家都是怎么实现的？

**Comment by czy88840616** [comment]
monorepo……

**Reply by xmsz** [reply]
感觉monorepo还是有点麻烦 适合同项目多包 
有没有再简单点的实现

就像https://github.com/go-kratos/beer-shop这样

**Comment by rccoder** [comment]
试试用 rush.js 管理 monorepo

https://rushjs.io/

**Comment by waitingsong** [comment]
lerna 管理。packages/ 下每个包单独发布

**Comment by xmsz** [comment]
2022年09月02日 记录一下

- app
- shared
- manage-app

根据https://stackoverflow.com/questions/60896829/monorepo-with-rootdirs-produces-unwanted-sudirectories-such-as-src-in-outdi 这个回答

意外的，app可以成功构建了，但是manage-app不行...  搞不明白

但至少说明通过tsConfig简单的配置还是可以实现的，不需要使用其他第三方工具来管理

**Reply by xmsz** [reply]
还有一种最粗暴的方式 就是把rootDir设置为"../" 然后把其他目录放到exclude里。这样的话虽然会构建整个项目并且dist目录结果不对，但是至少可以构建成功。剩下就是把dist/xxx里文件拷贝出来

**Reply by xmsz** [reply]
吐血 ，由于rootDir更改了，dist目录结构也改了，需要很傻的这样处理
```
mv dist dist1 && mv dist1/src dist && rm -rf ./dist1
```

**Reply by waitingsong** [reply]
与其折腾hack，不如花点精力实现 mono 方案（比如 lerna），然后做成模板一了百了

**Reply by czy88840616** [reply]
或者npm 的work space方案

**Comment by StringKe** [comment]
https://nx.dev/

https://turbo.build/repo

monorepo 是比较方便的解决方案了

**Reply by xmsz** [reply]
好 我试一下 我看好像可以支持内部包 不用上传到npm 应该能满足需求
